### ISSUE TITLE
WWT Notebooks failing under JupyterLab 2

### ISSUE THREAD

**Issue by hickst** [WorldWideTelescope/pywwt#272]
I just fired up TLJH (https://github.com/jupyterhub/the-littlest-jupyterhub) and it installed JupyterLab 2.2.5.

All of my previous analysis notebooks are working but WWT notebooks are failing, at the step where the WWT viewer is created, with the terse message: `Error displaying widget: model not found`. Below is a screenshot from running the WWT `First Steps` notebook. I'm wondering if pywwt needs an update to work with JL 2?

![WWT-first-steps-error](https://user-images.githubusercontent.com/22797/92184821-038d6000-ee07-11ea-90b0-49756177a714.png)

**Comment by pkgw** [comment]
Thanks for reporting! I know that the current code should work with JLab 2, as per [the merged PR #260](https://github.com/WorldWideTelescope/pywwt/pull/260). What's probably going on here instead is that the various Jupyter extensions need to be set up correctly, and that can't happen automatically with a `pip install`. If the extensions aren't right, you'll get messages like those you're seeing. The [install instructions] give some info about what to do. You might also want to check out the BinderHub setup scripts for the pywwt-notebooks repo, which can be launched in mybinder.org, [here]

[install instructions]: https://pywwt.readthedocs.io/en/stable/installation.html
[here]: https://github.com/WorldWideTelescope/pywwt-notebooks/tree/master/binder

I'm going to close this under the presumption that the issue is something related to the above, but please do reopen if this seems to be a different issue that calls for code changes to the pywwt repo.

**Comment by hickst** [comment]
Thanks Peter! I will follow your leads and let you know.

**Comment by hickst** [comment]
I've been exploring this but having no luck getting a working version under Jupyter Lab version 2.2.6.

I'm using miniconda to build a separate environment in which to run, as follows:
```
conda create -n wwt
conda activate wwt
conda install seaborn
conda install -c conda-forge ipywidgets
conda install -c astropy astroquery
conda install -c wwt pywwt
conda install "nodejs>=10.0"
jupyter labextension install @jupyter-widgets/jupyterlab-manager ipyevents --no-build
pip install pyvo
jupyter lab build
```
From the references you provided, it is my understanding that this should be sufficient. However, whenever I load a WWT notebook I get the error described above: 'Error displaying widget: model not found'

I noticed that using the default `pywwt` install from the `wwt` channel gave me version `0.7.0` of pywwt, so I also tried installing a newer version:
```
pip install git+https://github.com/WorldWideTelescope/pywwt.git@5fc62ef
jupyter lab build
```
Same error. Next, I tried installing the widgetsnbextension separately into my Conda `base`:
```
conda install -n base -c conda-forge widgetsnbextension
jupyter lab build
```
Same error. Finally, I tried installing PyQt5 (just in case):
```
pip install PyQt5 "astroquery>=0.3.9" "PyYAML>=3.1.3"
jupyter lab build
```
Same error.

Note that my other astro notebooks still run. The failure I'm getting is in (WWT) notebooks which create and launch a WWTJupyterWidget in a cell:
```
wwt = WWTJupyterWidget()
wwt
```
This error is pretty terse. Are there are any error logs or other outputs I should be looking at? (I've seen nothing on the JL console). Any other installation steps that I should try?

**Comment by hickst** [comment]
Hi Peter,

I finally got time to get back to this and wondered if maybe this was a mis-match of the labmanager version. So, I installed in a new Conda environment as follows:
```
conda create -n wwt
conda activate wwt
conda install seaborn
conda install -c astropy astroquery
conda install -c conda-forge ipywidgets
conda install -c wwt pywwt
conda install "nodejs>=10.0"
pip install pyvo
jupyter labextension install @jupyter-widgets/jupyterlab-manager@2.0
```

Unfortunately, I'm still getting the same cryptic error:  "Error displaying widget: model not found".

This error is pretty terse. Are there are any error logs or other outputs I should be looking at? (I've seen nothing on the JL console). Any other installation steps that I should try?

**Comment by pkgw** [comment]
Hi Tom,

Sorry for losing track of this — I hit a "dismiss all" button on the GitHub notification screen and lost a bunch of to-dos ...

I think I see a problem here. We're no longer updating the separate `wwt` Anaconda channel regularly, so if you're getting your packages from there you'll be getting out-of-date code. My personal recommendation is to get everything through conda-forge. The following script works for me:

```
conda create -c conda-forge -n wwt astroquery ipywidgets nodejs pip pycairo python pywwt seaborn setuptools
conda activate wwt
jupyter labextension install @jupyter-widgets/jupyterlab-manager@2.0

# diagnostics:
jupyter nbextension list
jupyter serverextension list
jupyter labextension list
```

Does that help?

**Comment by hickst** [comment]
Well, that helped but there's still something going wrong. I started by trying to run the 'NASA Explanet Archive.ipynb' in Chrome (version 85.0.4183.121). As soon as the wwt widget is created the JupyterLab goes into "overdrive": several CPUs spike to very high loads, the whole JL GUI becomes extremely slow, the Busy/Idle indicator at the bottom flashes between the two states constantly, and the Saving Completed indicator at the bottom flashes continuously. This happens in other WWT notebooks, but sometimes only after an error in a cell. It's as if an error causes the notebook to do some kind of repetitive interaction with the kernel that bogs the system down.

I'm having a hard time pinning down where the problem is: I've tested on both Miniconda on Ubuntu 18.04 and Anaconda 3 on macOS 10.14.6 with same results in Chrome. Then I tested with Safari and couldn't reproduce the problem. I'm assuming from the Windows notebook names that you are using an MS browser and not seeing these problems?

**Comment by hickst** [comment]
Hi Peter,

I forgot to mention that, if you are not updating the `wwt` regularily (or at all?), you might want to update the main installation documentation page: https://pywwt.readthedocs.io/en/stable/installation.html to not use the `wwt` channel.

**Comment by hickst** [comment]
OK, more data points: I just tried a fresh install of Miniconda on an Ubuntu 18.04 laptop, then installed WWT/JupyterLab as you recommended above. Checked out pywwt notebooks from GitHub. Started JupyterLab. Browsed to it with latest Chrome and tried the 'Visualizing Imagery.ipynb' notebook. Unfortunately, after the `wwt` widget was created, I saw the same issues as I previously reported: several CPUs went to %75+, constant blinking of JL's Busy/Idle and Saving Completed indicators and extremely slow response to mouse clicks, drags, movements, etc. Shut down the notebook, kernel, and JL server. Restarted the JL server, browsed to it with the latest Firefox for Linux, opened the Visualizing Imagery notebook and got the same issues with Firefox as I previously saw in Chrome. Shutdown the notebook and kernel and all CPUs went back to ~3%.
Not sure what's going on but I hope this helps.

**Comment by pkgw** [comment]
Some of the constant activity will be a combination of the inherent nature of WWT, which is like a video-game renderer and so always keeps one CPU busy, and a known issue with the pywwt Jupyter framework, #258. That being said, on my machine responsiveness is typically just fine as-is.

I'm hoping to fix #258 soon, but currently fighting with updating our CI to get it consistently passing with Python 3.8 (#275).

**Comment by hickst** [comment]
Thanks for the explanation. Like I said, the lag is minimal under Safari but debilitating under Chrome and Firefly: e.g., I had to wait about 5-8 seconds just to stretch a window frame.
Good luck with the CI.

**Comment by pkgw** [comment]
Hi Tom — I've worked through the issues I mentioned above. If you [try pywwt 0.10.1](https://mybinder.org/v2/gh/WorldWideTelescope/pywwt-notebooks/master?urlpath=lab/tree/Start%20Here.ipynb), how different are things? The Jupyter busy indicator should be a lot less active, at a minimum.

**Comment by hickst** [comment]
Hi Peter,

Thanks for the fixes. I tested all the non-workshop WWT notebooks and all appear to be working fine now under Miniconda 3 on both Ubuntu 8.0.4 and macOS 10.14.6. It really is night and day compared to the version with the issue: I don't think I saw any lag on either machine (and the connection to the Linux machine was tunneled). Here are the details of the successful JL installation, after installing Miniconda 3:
```
conda create -n wwt -c conda-forge astroquery ipywidgets "nodejs>=10.0" pip pycairo python pyYAML pywwt seaborn setuptools
conda activate wwt
pip install pyvo
jupyter labextension install @jupyter-widgets/jupyterlab-manager@2.0
```
Note that I had to specify `nodejs` >= 10.0, as version 6 was getting installed by default and `jupyterlab-manager` complained and wouldn't work with that version.

Note that `pyvo` was strictly for my use -- I don't think any of the notebooks require it.

I did have to add `pyYAML` to the list, specifically for the `GRBs Over Time.ipynb` notebook.
```
